lib/lf/assembly/assembler.h

```diff
@@ -0,0 +1,219 @@
+ * @tparam ASSEMBLER a type providing the computation of element matrices
+ * @param dof_handler_trial a dof handler object for column space @see
+ * DofHandler
+ * @param dof_handler_test a dof handler object for row space @see DofHandler
+ * @param assembler assembler object for passing all kinds of data
+ * @param matrix matrix object to which the assembled matrix will be added.
+ *               The matrix object is not set to zero in the beginning!
+ *
+ * This method performs cell-oriented assembly controlled by a local-to-global
+ * index map ("dof handler").
+ *
+ * #### type requirements of template arguments
+ *
+ * - TMPMATRIX is a rudimentary matrix type and must
+ * + provide a constructor taking two matrix dimension arguments
+ * + have a method `AddtoEntry(i,j,value_to_add)` for adding to a matrix entry
+ * + be copyable and assignable
+ * A model type is COOMatrix.
+ * - ASSEMBLER is a type capable of local assembly of element matrices. It must
+ * + have an `Eval()` method returning the element matrix for a cell
+ * + supply an `isActive()` method for selecting cells to be taken into account
+ * in assembly
+ * + provide a matrix type `ElemMat` for objects containing the element
+ * matrices.
+ */
+template <int CODIM, typename TMPMATRIX, class ASSEMBLER>
+void AssembleMatrixLocally(const DofHandler &dof_handler_trial,
+                           const DofHandler &dof_handler_test,
+                           ASSEMBLER &assembler, TMPMATRIX &matrix) {
+  // Type of matrix entries, usually either double or complex.
+  using scalar_t = typename TMPMATRIX::Scalar;
+  // Type for element matrix
+  using elem_mat_t = typename ASSEMBLER::ElemMat;
+  // Underlying mesh
+  const lf::mesh::Mesh &mesh(dof_handler_trial.getMesh());
+
+  LF_ASSERT_MSG(&mesh == &dof_handler_test.getMesh(),
+                "Trial and test space must be defined on the same mesh");
+
+  // Central assembly loop over entities of co-dimension specified by
+  // the template argument CODIM
+  for (const lf::mesh::Entity &entity : mesh.Entities(CODIM)) {
+    // Some entities may be skipped
+    if (assembler.isActive(entity)) {
+      // Size, aka number of rows and columns, of element matrix
+      const size_type nrows_loc = dof_handler_test.GetNoLocalDofs(entity);
+      const size_type ncols_loc = dof_handler_trial.GetNoLocalDofs(entity);
+      // row indices of for contributions of cells
+      lf::base::RandomAccessRange<const gdof_idx_t> row_idx(
+          dof_handler_test.GlobalDofIndices(entity));
+      // Column indices of for contributions of cells
+      lf::base::RandomAccessRange<const gdof_idx_t> col_idx(
+          dof_handler_trial.GlobalDofIndices(entity));
+      // Request local matrix from assembler object. In the case CODIM = 0,
+      // when `entity` is a cell, this is the element matrix
+      const elem_mat_t elem_mat(assembler.Eval(entity));
+      LF_ASSERT_MSG(elem_mat.rows() >= nrows_loc,
```

Comment:
Hä, maybe I'm too stupid, but can we not require `elem_mat.rows() == nrows_loc`? What about the extra rows/columns of the `elem_mat` if `elem_mat.rows() > nrows_loc`?

Reply:
A user could pre-allocated a matrix of maximum size and for most cells use only parts of it, for instance in an hp-FE context.
Use case: `Eigen::Matrix4d` for linear FEM and both triangles and quads.

Reply:
Ah I see. Hmm I'm not sure that this is the right approach to handle quadrilaterals.
Advantages of current approach:
- Computation Speed for "small" element matrices: fixed size matrices are allocated on the stack + 4x4 matrices are not much bigger than 3x3 matrices.

Disadvantages of current approach:
- Using higher order basis functions the overhead for quadrilaterals increases rapidly: 9x9 matrices are much bigger than 6x6 matrices. In this case it would probably make more sense to return appropriately size dynamic matrices.
- Errors may not be recognized, e.g. if `elemt_mat.rows() > nrow_loc` is really an error.
- Not very intuitive, returning a padded 4x4 matrix that is actually a 3x3 matrix is somwhat obscure.

Reply:
Gives me some headache, look at `assembly_tests.cc`, where the assembler defines a fixed-size matrix data type.
Of course, everything could be made dynamic, but since this is an innermost-loop operation efficiency should be a concern. Don't change it for the moment; I will think about it.

Reply:
Should we make an issue out of this?